### PR TITLE
Feat/awicm3.2

### PR DESCRIPTION
**Pull request by JanStreffing**
This merge brings AWI-CM3 v3.2 (https://awi-cm3-documentation.readthedocs.io/en/latest/releases.html#awi-cm3-v3-2) to the esm_tools release branch

### CODE REVIEW

File: configs/setups/awicm3/awicm3.yaml [hunk 1]
```diff
@@ -209,6 +237,13 @@ oifs:
                                                 ORBMODE: "fixed_year"
                                                 ORBIY: ${start_date!syear}
 
+        choose_general.version:
+                "v3.2":
+                        add_namelist_changes:
+                                fort.4:
+                                        NAERAD:
+                                                NAERANT_SCALE: "1"
+
```
Comment (mandresm) [inline configs/setups/awicm3/awicm3.yaml]: Isn't this something that should be in the `major_version` switch?

File: configs/setups/awicm3/awicm3.yaml [hunk 2]
```diff
@@ -413,10 +445,51 @@ fesom:
...
+                                        tracer_phys:
+                                                surf_relax_s: "${surf_relax_s}"
+                                                use_momix: false
+                                                kv0_const: false
+                                namelist.ice:
+                                        ice_therm:
+                                                albsn: 0.81
+                                                albsnm: 0.77
+                                                albi: 0.6
+                                                albim: 0.43
```
Comment (mandresm) [inline configs/setups/awicm3/awicm3.yaml]: Does it make sense to put all this under the `major_version` switch?

File: configs/setups/awicm3/awicm3.yaml [hunk 3]
```diff
@@ -577,15 +650,52 @@ oasis3mct:
...
+                                rstas.nc:
+                                        - 'heat_oce:heat_swo <--gauswgt_gss-- A_Qns_oce:A_Qs_all'
+                                        - 'prec_oce:snow_oce:evap_oce:subl_oce <--gauswgt_gss-- A_Precip_liquid:A_Precip_solid:A_Evap:A_Subl'
+                                        - 'heat_ico <--gauswgt_gss-- A_Q_ice'
+                                        - 'taux_oce:tauy_oce:taux_ico:tauy_ico <--bicubic-- A_TauX_oce:A_TauY_oce:A_TauX_ice:A_TauY_ice'
+                                        - 'R_Runoff_atm:R_Calving_atm <--gauswgt_gss-- A_Runoff:A_Calving'
+                                        - 'hydr_oce:enth_oce <--gauswgt_gss-- R_Runoff_oce:R_Calving_oce'
+                                rstos.nc:
+                                        - 'A_SST:A_Ice_frac:A_Snow_thickness:A_Ice_temp:A_Ice_albedo <--gauswgt_i-- sst_feom:sie_feom:snt_feom:ist_feom:sia_feom'
+                                        - 'A_CurX:A_CurY <--gauswgt_i-- u_feom:v_feom'
```
Comment (mandresm) [inline configs/setups/awicm3/awicm3.yaml]: Why adding this, when most of the versions, except v3.2 use the same?

File: configs/setups/awicm3/awicm3.yaml [hunk 4]
```diff
@@ -577,15 +650,52 @@ oasis3mct:
...
```
Comment (mandresm) [inline configs/setups/awicm3/awicm3.yaml:658]: Why removing the first lines and adding the later ones, when most of the versions, except v3.2 use the same?

File: configs/setups/awicm3/awicm3.yaml [hunk 5]
```diff
@@ -577,15 +650,52 @@ oasis3mct:
...
+                        coupling_target_fields:
+                                rstas.nc:
+                                        - 'heat_oce:heat_swo <--gauswgt_c-- A_Qns_oce:A_Qs_all'
+                                        - 'prec_oce:snow_oce:evap_oce:subl_oce <--gauswgt_c-- A_Precip_liquid:A_Precip_solid:A_Evap:A_Subl'
+                                        - 'heat_ico <--gauswgt_gss-- A_Q_ice'
+                                        - 'taux_oce:tauy_oce:taux_ico:tauy_ico <--bicubic-- A_TauX_oce:A_TauY_oce:A_TauX_ice:A_TauY_ice'
+                                        - 'R_Runoff_atm:R_Calving_atm <--gauswgt_c-- A_Runoff:A_Calving'
+                                        - 'hydr_oce:enth_oce <--gauswgt_c-- R_Runoff_oce:R_Calving_oce'
+                                rstos.nc:
+                                        - 'A_SST:A_Ice_frac:A_Snow_thickness:A_Ice_temp:A_Ice_albedo <--gauswgt_i-- sst_feom:sie_feom:snt_feom:ist_feom:sia_feom'
```
Comment (mandresm) [inline configs/setups/awicm3/awicm3.yaml:675]: This is the only one that is actually different, if you keep it like it is right now, you can revert all the other changes that duplicate a lot of information for the `coupling_target_fields`

File: configs/setups/awicm3/awicm3.yaml [hunk 6]
```diff
@@ -594,25 +704,33 @@ oasis3mct:
...
+                        coupling_target_fields:
+                                rstas.nc:
+                                        - 'heat_oce:heat_swo <--gauswgt_gss-- A_Qns_oce:A_Qs_all'
+                                        - 'prec_oce:snow_oce:evap_oce:subl_oce <--gauswgt_gss-- A_Precip_liquid:A_Precip_solid:A_Evap:A_Subl'
+                                        - 'heat_ico <--gauswgt_gss-- A_Q_ice'
+                                        - 'taux_oce:tauy_oce:taux_ico:tauy_ico <--bicubic-- A_TauX_oce:A_TauY_oce:A_TauX_ice:A_TauY_ice'
+                                        - 'R_Runoff_atm:R_Calving_atm <--gauswgt_gss-- A_Runoff:A_Calving'
+                                        - 'hydr_oce:enth_oce <--gauswgt_gss-- R_Runoff_oce:R_Calving_oce'
+                                rstos.nc:
+                                        - 'A_SST:A_Ice_frac:A_Snow_thickness:A_Ice_temp:A_Ice_albedo <--gauswgt_i-- sst_feom:sie_feom:snt_feom:ist_feom:sia_feom'
```
Comment (mandresm) [inline configs/setups/awicm3/awicm3.yaml:693]: Why removing the first lines and adding the later ones, when most of the versions, except v3.2 use the same?

File: configs/setups/awicm3/awicm3.yaml [hunk 7]
```diff
@@ -594,25 +704,33 @@ oasis3mct:
...
```
Comment (mandresm) [inline configs/setups/awicm3/awicm3.yaml:710]: As mentioned above, I suggest to revert these changes.

**Review by mandresm — CHANGES_REQUESTED** [review]
Hi @JanStreffing, nice PR!

I have some comments about information duplication in `awicm3.yaml`. Other than that there are a bunch of new namelists in `namelists/fesom2/2.0/awicm3/v3.2` and another set in `namelists/fesom2/2.5/awicm3/v3.2`. Is this intentional? I would think that you just wanted the ones of fesom 2.5.

File: configs/setups/awicm3/awicm3.yaml (hunk 1 above)
Comment (JanStreffing) [reply configs/setups/awicm3/awicm3.yaml]: yes, ty! moved

File: configs/setups/awicm3/awicm3.yaml (hunk 2 above)
Comment (JanStreffing) [reply configs/setups/awicm3/awicm3.yaml]: yes, i also added moving for cvmix namelist, in case someone wants to use it. Also move the default namelist to `${esm_namelist_dir}/fesom2/2.5/awicm3/v3.2/`

Comment (JanStreffing) [reply configs/setups/awicm3/awicm3.yaml]: this needs testing on my part. back to draft for now.

File: configs/setups/awicm3/awicm3.yaml (hunk 3 above)
Comment (JanStreffing) [reply configs/setups/awicm3/awicm3.yaml]: controlled per major version. There is no real default here, even if we are now for the first time changing. I don't think the old stuff should be the default. Thus new def per major version.

File: configs/setups/awicm3/awicm3.yaml (hunk 4 above)
Comment (JanStreffing) [reply configs/setups/awicm3/awicm3.yaml:658]: see above. IMO this is not a case of default and overwrite.

**Comment by JanStreffing** [comment]
It seems I have messed something up with my last 2 commits:
```

python -mpdb ~/.local/bin/esm_runscripts awicm3-v3.2-levante-TCO95L91-CORE2.yaml -e tuningtest2 --open-run -c
> /home/a/a270092/.local/bin/esm_runscripts(3)<module>()
-> import re
(Pdb) c
Traceback (most recent call last):
  File "/sw/spack-levante/mambaforge-4.11.0-0-Linux-x86_64-sobz6z/lib/python3.9/pdb.py", line 1723, in main
    pdb._runscript(mainpyfile)
  File "/sw/spack-levante/mambaforge-4.11.0-0-Linux-x86_64-sobz6z/lib/python3.9/pdb.py", line 1583, in _runscript
    self.run(statement)
  File "/sw/spack-levante/mambaforge-4.11.0-0-Linux-x86_64-sobz6z/lib/python3.9/bdb.py", line 580, in run
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/home/a/a270092/.local/bin/esm_runscripts", line 3, in <module>
    import re
  File "/home/a/a270092/esm_tools/src/esm_runscripts/cli.py", line 281, in main
    Setup = SimulationSetup(command_line_config)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/sim_objects.py", line 42, in __init__
    self.config = prepare.run_job(self.config)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/prepare.py", line 17, in run_job
    helpers.evaluate(config, "prepare", "prepare_recipe")
  File "/home/a/a270092/esm_tools/src/esm_runscripts/helpers.py", line 69, in evaluate
    config = esm_plugin_manager.work_through_recipe(
  File "/home/a/a270092/esm_tools/src/esm_plugin_manager/esm_plugin_manager.py", line 141, in work_through_recipe
    config = getattr(submodule, workitem)(config)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/batch_system.py", line 685, in find_openmp
    config2 = prepare.resolve_some_choose_blocks(config2)
  File "/home/a/a270092/esm_tools/src/esm_runscripts/prepare.py", line 126, in resolve_some_choose_blocks
    esm_parser.choose_blocks(config, blackdict=config._blackdict)
  File "/home/a/a270092/esm_tools/src/esm_parser/esm_parser.py", line 2600, in choose_blocks
    resolve_basic_choose(
  File "/home/a/a270092/esm_tools/src/esm_parser/esm_parser.py", line 1477, in resolve_basic_choose
    for update_key, update_value in choices_available[choice].items():
AttributeError: 'NoneType' object has no attribute 'items'
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program
> /home/a/a270092/esm_tools/src/esm_parser/esm_parser.py(1477)resolve_basic_choose()
-> for update_key, update_value in choices_available[choice].items():
(Pdb) update_key
*** NameError: name 'update_key' is not defined
(Pdb) choice
'v3.2'
(Pdb) choices_available[choice]
(Pdb) choices_available[choice].items()
*** AttributeError: 'NoneType' object has no attribute 'items'
```

Perhaps you have sharper eyes than me? When happened?

**Comment by JanStreffing** [comment]
Question:
I'm trying to use the new fesom2.5.yaml. like this:

in awicm3.yaml:
```
fesom:
        version: "2.0"
        choose_general.major_version:
                "v3.2":
                        add_version: "2.5"
```

Is this the right way? In the finished yaml I still get:

```
    debug_info:
        loaded_from_file: /home/a/a270092/esm_tools/configs//components/fesom/fesom-2.0.yaml
    destination: fesom-2.0

```

**Comment by mandresm** [comment]
> Question: I'm trying to use the new fesom2.5.yaml. like this:
> 
> in awicm3.yaml:
> 
> ```
> fesom:
>         version: "2.0"
>         choose_general.major_version:
>                 "v3.2":
>                         add_version: "2.5"
> ```
> 
> Is this the right way? In the finished yaml I still get:
> 
> ```
>     debug_info:
>         loaded_from_file: /home/a/a270092/esm_tools/configs//components/fesom/fesom-2.0.yaml
>     destination: fesom-2.0
> ```

I don't think so, in principle, it should be possible to run it without the add, but there is a problem there, and it's being there for a while (sorry...). What I've tipically done in this case is to include all the versions in the choose and not define one default one. But this might not work either, at least for esm_master. I'll try to have a look at a solution this afternoon.

**Comment by mandresm** [comment]
I can confirm that there is a bug here. I am half way of finding out what's going on, but I'm in a conference, so I'll be slower than usual.

**Comment by JanStreffing** [comment]
Thanks for digging!

**Comment by mandresm** [comment]
Hi @JanStreffing,

I have managed to fix the problems with the version switching... the functions affected by the bugs were written by me on my early days of ESM-Tools :)

You'll find the fixes in the branch `fix/recursive_resolution_of_early_variables_and_fix_find_key`. You can just merge that branch here (just departed from `release`) or cherry-pick the single commit `3028599d2a27fdec65cba2e4c0f498eaafdec544` because I don't think the PR (#1063) will be reviewed earlier than Wednesday.

The failed tests in #1063 shouldn't be a problem.

**Comment by JanStreffing** [comment]
I think i have to do some more tests, now that we actually use the new fesom2 yaml.

**Comment by JanStreffing** [comment]
Thank you for the work on the fix @mandresm. I'll give this a try over the next few days.

**Comment by JanStreffing** [comment]
I just merged release after #1063 into https://github.com/esm-tools/esm_tools/tree/feat/awicm3.2 and did a check run.

In awicm3.yaml I set:

```
general:
        choose_version:
                v3.2:
                        major_version: v3.2

fesom:
        version: "2.0"
        choose_general.major_version:
                "v3.2":
                        add_version: "2.5"
```

In the runscript:
```
general:
        version: "v3.2"
```

Nevertheless, the finished yaml has:

```
general:
        major_version: v3.2

fesom:
        version: 2
```

The latter should now be 2.5. And indeed the options from the 2.5 yaml are not read. Any mistake on my part?

**Comment by JanStreffing** [comment]
Checkrun on levante: 
/work/ab0246/a270092/runtime/awicm3-v3.2/25test_1

**Comment by JanStreffing** [comment]
ready to merge :beers:

**Comment by JanStreffing** [comment]
Halt! one more fix...

**Comment by JanStreffing** [comment]
done

File: configs/setups/awicm3/awicm3.yaml (hunk 3 above)
Comment (mandresm) [reply configs/setups/awicm3/awicm3.yaml]: In my view it's not a matter of default but a matter spotting the difference between versions right away. Anyway, just a suggestion. I'll approve it as it is.

**Comment by mandresm** [comment]
Are you aware of this changes in the namcouple of the frontiers version?

https://github.com/esm-tools/esm_tools/actions/runs/6572115118/job/17852600385?pr=1056#step:8:395

**Comment by JanStreffing** [comment]
I was not! I'll fix that.

**Comment by mandresm** [comment]
I'm adding a test runscript to the CI for the 3.2 version and doing a real test in Levante. Should I perform the real test somwhere else? (that won't delay the merge, and I'll do it between today and tomorrow)

**Comment by JanStreffing** [comment]
i think it's a good idea to do a test on juwels as well. We might not have albedo yet, correct?

**Comment by mandresm** [comment]
> i think it's a good idea to do a test on juwels as well. We might not have albedo yet, correct?

Albedo is something for next week. I'm running the tests on Levante. I'll do the same on Juwels now that all is configured in esm_tests. Tomorrow I can do the same in Aleph.

**Comment by JanStreffing** [comment]
great :+1: :

**Comment by mandresm** [comment]
#approve-changes

**Comment by mandresm** [comment]
#bump

**Comment by mandresm** [comment]
#bump please...

**Comment by mandresm** [comment]
#bump sorry Jan... something change in GitHub actions and broke our bump workflow...

**Comment by mandresm** [comment]
#bump

**Review by mandresm — APPROVED** [review]
